dvc/repo/experiments/queue/base.py

```diff
@@ -53,22 +53,22 @@ class QueueEntry:
     branch: Optional[str]
     name: Optional[str]
 
-    def __eq__(self, other: object):
-        return (
-            isinstance(other, QueueEntry)
-            and self.dvc_root == other.dvc_root
-            and self.scm_root == other.scm_root
-            and self.stash_ref == other.stash_ref
-            and self.stash_rev == other.stash_rev
-        )
-
     def asdict(self) -> Dict[str, Any]:
         return asdict(self)
 
     @classmethod
     def from_dict(cls, d: Dict[str, Any]) -> "QueueEntry":
         return cls(**d)
 
+    def _key(self):
+        return (self.dvc_root, self.scm_root, self.stash_ref, self.stash_rev)
+
+    def __hash__(self):
+        return hash(self._key())
+
+    def __eq__(self, other):
+        return isinstance(other, QueueEntry) and self._key() == other._key()
+
```

Comment:
This should be implemented using the automatically generated hash and comparison features in `dataclass`: https://docs.python.org/3/library/dataclasses.html

So for optional fields we would just set them like

```py
from dataclasses import dataclass, field

@dataclass
class QueueEntry:
    ...
    branch: Optional[str] = field(compare=False)
````